### PR TITLE
[Bugfix]: Item prices does not work #500

### PR DESCRIPTION
**Pull request by tinutac**
My idea is to use the strategy below for updating the average price. Then you get a quite good approximation of a FIFO approach of stock use.
Basically: 
- positive stock entry  -> update average price (computed as a (sum of added stock * individual price)/ total number of items).
- negative stock entry:
 - if  negative total, just set everything on zero (avg price, ~~total items~~)
 - if positive total:
   - smaller than the last positive stock entry use the price of the last positive stock entry as avg price. 
    - higher than the last positive stock just update the avg price using the same strategy as for a positive stock entry.

### CODE REVIEW

**Comment by tinutac** [comment]
@felicitus Hm... Any idea why the Travis CI build failed?

**Comment by felicitus** [comment]
It seems that your modification messes with the stock levels somehow, the unit test in question is here: 

https://github.com/partkeepr/PartKeepr/blob/master/src/PartKeepr/PartBundle/Tests/StockTest.php#L115

[Here's a doc](https://wiki.partkeepr.org/wiki/Developers/Unit_Testing) on how to run the unit tests on your machine. You can pass the filename of the individual test for debugging.

**Comment by tinutac** [comment]
I'm a little bit puzzled.
`$this->assertEquals($oldStockLevel` - 7, $newStockLevel);
I don't understand this line, actually. Where is that 7 coming from? I'm missing something?

**Comment by felicitus** [comment]
The unit tests removes 7 items from the stock and tests that the stock level is -7 then. It seems that your code prevents the change from happening. Yes, negative stock levels are allowed

**Comment by tinutac** [comment]
Ah, that explains it. As I've automatically corrected for negative stock :)
Ok. I'll take a look at it asap.

**Comment by tinutac** [comment]
Stupid question, probably, do you have a test database?

**Comment by felicitus** [comment]
Yes, and the reason is actually pretty simple:

Let's say you got a new batch of 501 screws, and made a mistake when entering them, so you actually only added 500 screws. At some point in time, the stock level shows 3 screws.

Now another user sees 4 screws and wants to remove them. If the system wouldn't allow negative stock, the user could only remove 3 screws, which is wrong.

File: src/PartKeepr/PartBundle/Entity/Part.php [hunk 1]
```diff
@@ -823,24 +823,40 @@ public function recomputeStockLevels()
...
+			if ($stockLevel->getStockLevel() > 0) {
+				$lastPosEntryQuant = $stockLevel->getStockLevel();
+				$lastPosEntryPrice = $stockLevel->getPrice();
+				$totalPartStockPrice += $lastPosEntryPrice * $lastPosEntryQuant;
+				$price = $totalPartStockPrice / $sum;
+			}
+			else {
+				if  ($sum < 0) {
+					$price = 0;
+					//$sum = 0;
```
Comment (felicitus) [inline src/PartKeepr/PartBundle/Entity/Part.php]: Please don't commit commented out lines

File: src/PartKeepr/PartBundle/Entity/Part.php [hunk 2]
```diff
@@ -823,24 +823,40 @@ public function recomputeStockLevels()
...
+					}
+					else {
+						$totalPartStockPrice += $stockLevel->getStockLevel() * $price;
+						$price = $totalPartStockPrice / $sum;
+					}
+			}
+		}
 
         $this->setStockLevel($sum);
+		$this->setAveragePrice($price);
```
Comment (felicitus) [inline src/PartKeepr/PartBundle/Entity/Part.php]: Indentation wrong, please follow [PSR-2](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md) for coding standards. You can use the [php-cs-fixer](http://cs.sensiolabs.org/) to fix the coding style.

Comment (tinutac) [inline src/PartKeepr/PartBundle/Entity/Part.php]: Should be ok now. My editor was simply on another coding standard :)

File: src/PartKeepr/PartBundle/Entity/Part.php (hunk 1 above)
Comment (tinutac) [inline src/PartKeepr/PartBundle/Entity/Part.php]: Removed it completely.

**Comment by tinutac** [comment]
Sorry, i was thinking from the "accounting" point of view (you can't use something you don't have in stock) but I follow your reasoning.

**Comment by tinutac** [comment]
I also made a quick fix (that I didn't commit/push yet) for that issue reported here: https://github.com/partkeepr/PartKeepr/issues/579#issuecomment-185803483 as I need at least to access my projects. If you want I can commit/push that one also here.

**Comment by tinutac** [comment]
Hi @felicitus, I've noticed yesterday a small bug still in the calculation. I had not committed it as I was still testing it on my system. 
Since we're allowing negative stock I had to modify a little bit the updating of the avg/stock value for a positive stock entry when having a total negative stock. 
I see that you've merged it already in the master of partkeepr. What's the best way to include this small mod also?

**Comment by felicitus** [comment]
Maybe this helps: http://stackoverflow.com/questions/12674304/github-reopening-a-merged-pull-request

If you provide the fix until tomorrow I can include it in the next version

**Comment by crazyimp** [comment]
Hi @tinutac what is the idea to zero the price for removing items from stock?
It would be great to see the price of the parts at the "time of removal". My use-case would be helping out friends with parts - given they got their own account (or by using the comment field), one could look at the reports to figure out who owes who (taking the current prices might be incorrect since the average price might have changed since).

**Comment by tinutac** [comment]
Hi @crazyimp. I'm not sure I understand your question correctly. You mean why when an item is removed from stock the price used is not the price of it when it was entered? We should include @felicitus in this discussion also as he's the one taking decisions about partkeepr. 
The setting of the price for the removal is a historic decision (prior to my involvement into this issue). If I'm correct, it seems you would like to couple the parts removal to a given user. I see a small issue with the price assigned to the removal. What do you do in this situation?
+ 5pieces 10$/piece (user1)
+ 5pieces 8$/piece (user2)
- 8pieces ???/piece (user3)
??? computed as average for those values? From what you say, what you would actually like to have it's the removal being split over a -5 x 10$ and a -2 x 8$ (we don't go in the details if you would not do the removal chronological as that starts another discussion).

**Comment by crazyimp** [comment]
@tinutac I'm only talking about the stock history, not necessarily related to stocking parts from different users.
Simple example:
Add 5x partXY, 1$ each. -- It shows up as Amount: 5, Price 1$.
Remove 1x -- History shows: Amount -1, Price: 0$. 
The 0$ price upon removal is what I'm talking about. Why doesn't it show up with the current price (average, we don't need the fifo part here ;))?

**Comment by tinutac** [comment]
Hi @crazyimp, as I said, that's a historical decision taken by @felicitus. I'm willing to take a look at the removal procedure/price (next week I'll have some more time as now I'm quiet caught up in another project) if he decides he would like to change this. 
I understood from your first example what you were talking about :). My question stays the same: what about removal of products from two batches (removal from a single batch is a quite straightforward example)? Or you're thinking about manual input of price for the used products? If you read in our previous conversations, partkeepr allows even negative stock so that's also a scenario that should be taken into account should it be decided to add prices to the stock removals.

**Pull request merged**